BLSQ/iaso
# IA-3365 user history backend

Explain what problem this PR is resolving

Related JIRA tickets : IA-3365

## Self proofreading checklist


## Doc
- Tell us where the doc can be found (docs folder, wiki, in the code...)

## Changes
- update `Modification` model:
    - use SET_NULL i.o DO_NOTHING for user foreign key
    - Add SQL migration to make existing logs consistent
- Add `ProfileAuditLogger, subclassing `AuditLogger`. The shape of the `new_value` and `past_value` entries matches the shape of existing logs for users and adds a few fields. It was kept flat for ease of use with existing front-end components.
- Add `ProfileAuditSerializer` and related nested serializers:
    - In the `new_value` field, we log a boolean `password_updated`.
    - On deletion, we log the time in a `deleted_at` field, in anticipation of soft delete feature (so the logs will be consistent)
- Add audit logging to 'create`method of `ProfileViewset `
- Refactor `partial_update` method of `ProfileViewset` to raise all potential exceptions before writing to the DB. Add audit logging when update is successful
- Refactor `update_single_profile_from_bulk` in the same way. Add  audit log on profile if update successful. Add audit log for teams if teams are updated
- Add audit logging to `BulkCreateUserFromCsvViewSet`.
- Add fixtures and tests


## How to test
- Login as super user to have full access to the `/api/logs/`
Go to iaso>admin>users
- create a single user
- Go to `/api/logs/contentType=iaso.profile&fields=past_value,new_value/`
- The latest log should be the user you created. It should have an empty `past_value`. `new_value.password_updated` should be `True`

- update existing user
- repeat above steps and check that values are correct

- select several users in the table
- With the speed dial, select update
- Update all possible fields and save
- Go to `/api/logs/contentType=iaso.profile&fields=past_value,new_value/`: you should find a log per user
- Go to `/api/logs/contentType=iaso.team&fields=past_value,new_value/`: you should find a log per user and per team modified

- click create from file button and confirm
- Go to `/api/logs/contentType=iaso.profile&fields=past_value,new_value/`: you should find a log per user


## Notes

I haven't refactored the code for user bulk create because we will go through it again for performance and move it in to a task shortly, so it can be done correctly then
